<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Background</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>The sommer package was developed to provide R users a powerful and reliable multivariate mixed model solver for different genetic and non-genetic analysis in diploid and polyploid organisms. This package allows the user to estimate variance components for a mixed model with the advantage of specifying the variance-covariance structure of the random effects, specify heterogeneous variances, and obtain other parameters such as BLUPs, BLUEs, residuals, fitted values, variances for fixed and random effects, etc. The core algorithms of the package are coded in C++ using the Armadillo library to opmitime dense matrix operations common in the derect-inversion algorithms. </p>

<p>The package is focused on problems of the type p &gt; n related to genomic prediction (hybrid prediction &amp; genomic selection) and GWAS analysis, although any general mixed model can be fitted as well. The package provides kernels to estimate additive (<code>A.mat</code>), dominance (<code>D.mat</code>), and epistatic (<code>E.mat</code>) relationship matrices that have been shown to increase prediction accuracy under certain scenarios or simply to estimate the variance components of such. The package provides flexibility to fit other genetic models such as full and half diallel models as well.</p>

<p>Vignettes aim to provide several examples in how to use the sommer package under different scenarios. We will spend the rest of the space providing examples for: </p>

<p>1) Heritability (\(h^2\)) calculation
2) Specifying heterogeneous variances in mixed models
3) Using the pin calculator
4) Half and full diallel designs (using the overlay)
5) Genomic selection (predicting mendelian sampling)
    + GBLUP
    + rrBLUP
6) Single cross prediction (hybrid prediction)
7) Spatial modeling (using the 2-dimensional splines)
8) Multivariate genetic models and genetic correlations
9) Final remarks</p>

<h2>Background</h2>

<p>The core of the package the <code>mmer</code>function which solve the mixed model equations. The functions are an interface to call the <code>NR</code> Direct-Inversion Newton-Raphson or Average Information (Tunnicliffe 1989; Gilmour et al. 1995; Lee et al. 2016). Since version 2.0 sommer can handle multivariate models. Following Maier et al. (2015), the multivariate (and by extension the univariate) mixed model implemented has the form:</p>

<p><br></p>

<p>\(y_1 = X_1\beta_1 + Z_1u_1 + \epsilon_1\) 
\(y_2 = X_2\beta_2 + Z_2u_2 + \epsilon_2\) 
&hellip;
\(y_i = X_i\beta_i + Z_iu_i + \epsilon_i\) </p>

<p><br></p>

<p>where \(y_i\) is a vector of trait phenotypes, \(\beta_i\) is a vector of fixed effects, \(u_i\) is a vector of random effects for individuals and \(e_i\) are residuals for trait ‘i’ (i = 1, …, t). The random effects (\(u_1\) &hellip; \(u_i\) and \(e_i\)) are assumed to be normally distributed with mean zero. X and Z are incidence matrices for fixed and random effects respectively. The distribution of the multivariate response and the phenotypic variance covariance (V) are:</p>

<p><br> </p>

<p>\(Y = X\beta + ZU + \epsilon_i\)</p>

<p><br></p>

<p>Y ~ MVN(\(X\beta\), V)</p>

<p><br></p>

<p>\[\mathbf{Y} = \left[\begin{array}
{r}
y_1 \
y_2 \
&hellip; \
y_t \
\end{array}\right]
\]</p>

<p><br></p>

<p>\[\mathbf{X} = \left[\begin{array}
{rrr}
X_1 & &hellip; & &hellip; \
\vdots & \ddots & \vdots\
&hellip; & &hellip; & X_t \
\end{array}\right]
\]</p>

<p><br></p>

<p>\[\mathbf{V} = \left[\begin{array}
{rrr}
Z_1 K{\sigma^2_{g_{1}}} Z_1' + H{\sigma^2_{\epsilon_{1}}} & &hellip; & Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}}\
 \vdots & \ddots & \vdots\
Z_1 K{\sigma_{g_{1,t}}} Z_t' + H{\sigma_{\epsilon_{1,t}}} & &hellip; & Z_t K{\sigma^2_{g_{t}}} Z_t' + H{\sigma^2_{\epsilon_{t}}} \
\end{array}\right]
\]</p>

<p><br></p>

<p>where K is the relationship or covariance matrix for the kth random effect (u=1,…,k), and R=I is an identity matrix for the residual term. The terms \(\sigma^2_{g_{i}}\) and \(\sigma^2_{\epsilon_{i}}\) denote the genetic (or any of the kth random terms) and residual variance of trait ‘i’, respectively and \(\sigma_{g_{_{ij}}}\) and \(\sigma_{\epsilon_{_{ij}}}\) the genetic (or any of the kth random terms) and residual covariance between traits ‘i’ and ‘j’ (i=1,…,t, and j=1,…,t). The algorithm implemented optimizes the log likelihood:</p>

<p><br></p>

<p>\(logL = &frac12; * ln(|V|) + ln(X'|V|X) + Y'PY\)</p>

<p><br></p>

<p>where || is the determinant of a matrix. And the REML estimates are updated using a Newton optimization algorithm of the form:</p>

<p><br></p>

<p>\(\theta^{k+1} = \theta^{k} + (H^{k})^{-1}*\frac{dL}{d\sigma^2_i}|\theta^k\)</p>

<p><br></p>

<p>Where, \(\theta\) is the vector of variance components for random effects and covariance components among traits, \(H^{-1}\) is the inverse of the Hessian matrix of second derivatives for the kth cycle, \(\frac{dL}{d\sigma^2_i}\) is the vector of first derivatives of the likelihood with respect to the variance-covariance components. The Eigen decomposition of the relationship matrix proposed by Lee and Van Der Werf (2016) was included in the Newton-Raphson algorithm to improve time efficiency. Additionally, the popular pin function to estimate standard errors for linear combinations of variance components (i.e. heritabilities and genetic correlations) was added to the package as well.</p>

<p>Please refer to the canonical papers listed in the Literature section to check how the algorithms work. We have tested widely the methods to make sure they provide the same solution when the likelihood behaves well but for complex problems they might lead to slightly different answers. If you have any concern please contact me at <a href="mailto:cova_ruber@live.com.mx">cova_ruber@live.com.mx</a>.</p>

<p>In the following section we will go in detail over several examples on how to use mixed models in univariate and multivariate case and their use in quantitative genetics.</p>

<p><br></p>

<h2>1) Marker and non-marker based heritability calculation</h2>

<p>The heritability is one of the most popular parameters among the breeding and genetics community because of the insight that provides in the inheritance of the trait. The heritability is usually estimated as narrow sense (\(h^2\); only additive variance in the numerator \(\sigma^2_A\)), and broad sense (\(H^2\); all genetic variance in the numerator \(\sigma^2_G\)).</p>

<p>In a classical breeding experiment with no molecular markers, special designs are performed to estimate and disect the additive (\(\sigma^2_A\)) and non-additive (i.e. dominance \(\sigma^2_D\)) variance along with environmental variability. Designs such as generation analysis, North Carolina designs are used to disect \(\sigma^2_A\) and \(\sigma^2_D\) to estimate the narrow sense heritability (\(h^2\)). When no special design is available we can still disect the genetic variance (\(\sigma^2_G\)) and estimate the broad sense heritability. In this first example we will show the broad sense estimation which doesn&#39;t use covariance structures for the genotipic effect (i.e. genomic or additive relationship matrices). For big models with no covariance structures, sommer&#39;s direct inversion is a bad idea to use but we will show anyways how to do it, but keep in mind that for very sparse models we recommend using the lmer function from the lme4 package or any other package using MME-based algorithms (i.e. asreml-R).</p>

<p>The following dataset has 41 potato lines evaluated in 5 locations across 3 years in an RCBD design. We show how to fit the model and extract the variance components to calculate the \(h^2\).</p>

<pre><code class="r">library(sommer)
data(DT_example)
DT &lt;- DT_example
A &lt;- A_example

ans1 &lt;- mmer(Yield~1,
             random= ~ Name + Env + Env:Name + Env:Block,
             rcov= ~ units,
             data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -40.765   21:49:50      0           0
##     2      -30.2657   21:49:50      0           0
##     3      -25.8227   21:49:50      0           1
##     4      -24.7277   21:49:50      0           1
##     5      -24.7203   21:49:51      1           1
##     6      -24.7202   21:49:51      1           1
</code></pre>

<pre><code class="r">summary(ans1)$varcomp
</code></pre>

<pre><code>##                         VarComp  VarCompSE    Zratio Constraint
## Name.Yield-Yield       3.718279  1.6959834 2.1924029   Positive
## Env.Yield-Yield       12.008450 12.2771178 0.9781164   Positive
## Env:Name.Yield-Yield   5.152643  1.4923912 3.4526091   Positive
## Env:Block.Yield-Yield  0.000000  0.1156675 0.0000000   Positive
## units.Yield-Yield      4.366189  0.6573086 6.6425245   Positive
</code></pre>

<pre><code class="r">(n.env &lt;- length(levels(DT$Env)))
</code></pre>

<pre><code>## [1] 3
</code></pre>

<pre><code class="r">pin(ans1, h2 ~ V1 / ( V1 + (V3/n.env) + (V5/(2*n.env)) ) )
</code></pre>

<pre><code>##     Estimate        SE
## h2 0.6032715 0.1344582
</code></pre>

<p>Recently with markers becoming cheaper, thousand of markers can be run in the breeding materials. When markers are available, an special design is not neccesary to disect the additive genetic variance. The availability of the additive, dominance and epistatic relationship matrices allow us to estimate \(\sigma^2_A\), \(\sigma^2_D\) and \(\sigma^2_I\), although given that A, D and E are not orthogonal the interpretation of models that fit more than A and D become cumbersome.</p>

<p>Assume you have a population (even unreplicated) in the field but in addition we have genetic markers. Now we can fit the model and estimate the genomic heritability that explains a portion of the additive genetic variance (with high marker density \(\sigma^2_A\) = \(\sigma^2_g\))</p>

<pre><code class="r">data(DT_cpdata)
DT &lt;- DT_cpdata
GT &lt;- GT_cpdata
MP &lt;- MP_cpdata
DT$idd &lt;-DT$id; DT$ide &lt;-DT$id
### look at the data
A &lt;- A.mat(GT) # additive relationship matrix
D &lt;- D.mat(GT) # dominance relationship matrix
E &lt;- E.mat(GT) # epistatic relationship matrix
ans.ADE &lt;- mmer(color~1, 
                 random=~vs(id,Gu=A) + vs(idd,Gu=D), 
                 rcov=~units,
                 data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -123   21:49:51      0           0
##     2      -107.864   21:49:52      1           0
##     3      -103.867   21:49:52      1           0
##     4      -103.315   21:49:52      1           0
##     5      -103.294   21:49:52      1           0
##     6      -103.293   21:49:52      1           0
</code></pre>

<pre><code class="r">(summary(ans.ADE)$varcomp)
</code></pre>

<pre><code>##                       VarComp    VarCompSE   Zratio Constraint
## u:id.color-color  0.003662202 0.0012194130 3.003250   Positive
## u:idd.color-color 0.001820079 0.0007406216 2.457502   Positive
## units.color-color 0.002106929 0.0002864724 7.354736   Positive
</code></pre>

<pre><code class="r">pin(ans.ADE, h2 ~ (V1) / ( V1+V3) )
</code></pre>

<pre><code>##     Estimate         SE
## h2 0.6347926 0.08840488
</code></pre>

<pre><code class="r">pin(ans.ADE, h2 ~ (V1+V2) / ( V1+V2+V3) )
</code></pre>

<pre><code>##     Estimate         SE
## h2 0.7223783 0.05563774
</code></pre>

<p>In the previous example we showed how to estimate the additive (\(\sigma^2_A\)), dominance (\(\sigma^2_D\)), and epistatic (\(\sigma^2_I\)) variance components based on markers and estimate broad (\(H^2\)) and narrow sense heritability (\(h^2\)). Notice that we used the <code>vs()</code> function which indicates that the random effect inside the parenthesis (i.e. id, idd or ide) has a covariance matrix (A, D, or E), that will be specified in the Gu argument  of the vs() function. Please DO NOT provide the inverse but the original covariance matrix.</p>

<h2>2) Specifying heterogeneous variances in univariate models</h2>

<p>Very often in multi-environment trials, the assumption that genetic variance is the same across locations may be too naive. Because of that, specifying a general genetic component and a location specific genetic variance is the way to go.</p>

<p>We estimate variance components for \(GCA_2\) and \(SCA\) specifying the variance structure.</p>

<pre><code class="r">data(DT_cornhybrids)
DT &lt;- DT_cornhybrids
DTi &lt;- DTi_cornhybrids
GT &lt;- GT_cornhybrids
### fit the model
modFD &lt;- mmer(Yield~1, 
               random=~ vs(at(Location,c(&quot;3&quot;,&quot;4&quot;)),GCA2), 
               rcov= ~ vs(ds(Location),units),
               data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -190.104   21:49:52      0           0
##     2      -171.543   21:49:52      0           0
##     3      -165.319   21:49:52      0           0
##     4      -164.691   21:49:52      0           0
##     5      -164.684   21:49:53      1           0
##     6      -164.684   21:49:53      1           0
</code></pre>

<pre><code class="r">summary(modFD)
</code></pre>

<pre><code>## ============================================================
##          Multivariate Linear Mixed Model fit by REML         
## **********************  sommer 4.0  ********************** 
## ============================================================
##          logLik      AIC      BIC Method Converge
## Value -164.6839 331.3677 335.3592     NR     TRUE
## ============================================================
## Variance-Covariance components:
##                     VarComp VarCompSE Zratio Constraint
## 3:GCA2.Yield-Yield    62.48     53.45  1.169   Positive
## 4:GCA2.Yield-Yield    97.99     79.56  1.232   Positive
## 1:units.Yield-Yield  216.82     30.77  7.047   Positive
## 2:units.Yield-Yield  216.82     30.77  7.047   Positive
## 3:units.Yield-Yield  493.05     77.27  6.381   Positive
## 4:units.Yield-Yield  711.98    111.63  6.378   Positive
## ============================================================
## Fixed effects:
##   Trait      Effect Estimate Std.Error t.value
## 1 Yield (Intercept)    138.1    0.9442   146.3
## ============================================================
## Groups and observations:
##        Yield
## 3:GCA2    20
## 4:GCA2    20
## ============================================================
## Use the &#39;$&#39; sign to access results and parameters
</code></pre>

<p>In the previous example we showed how the <code>at()</code> function is used in the <code>mmer</code> solver. By using the <code>at</code> function you can specify that i.e. the GCA2 has a different variance in different Locations, in this case locations 3 and 4, but also a main GCA variance. This is considered a CS + DIAG (compound symmetry + diagonal) model.</p>

<p>In addition, other functions can be added on top to fit models with covariance structures, i.e. the Gu argument from the <code>vs()</code> function to indicate a covariance matrix (A, pedigree or genomic relationship matrix)</p>

<pre><code class="r">data(DT_cornhybrids)
DT &lt;- DT_cornhybrids
DTi &lt;- DTi_cornhybrids
GT &lt;- GT_cornhybrids
GT[1:4,1:4]
</code></pre>

<pre><code>##             A258       A634        A641        A680
## A258  2.23285528 -0.3504778 -0.04756856 -0.32239362
## A634 -0.35047780  1.4529169  0.45203869 -0.02293680
## A641 -0.04756856  0.4520387  1.96940221 -0.09896791
## A680 -0.32239362 -0.0229368 -0.09896791  1.65221984
</code></pre>

<pre><code class="r">### fit the model
modFD &lt;- mmer(Yield~1, 
              random=~ vs(at(Location,c(&quot;3&quot;,&quot;4&quot;)),GCA2,Gu=GT), 
              rcov= ~ vs(ds(Location),units),
              data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -191.286   21:49:53      0           0
##     2      -172.247   21:49:53      0           0
##     3      -165.948   21:49:53      0           0
##     4      -165.248   21:49:53      0           0
##     5      -165.23   21:49:53      0           0
##     6      -165.229   21:49:53      0           0
##     7      -165.229   21:49:53      0           0
</code></pre>

<pre><code class="r">summary(modFD)
</code></pre>

<pre><code>## ============================================================
##          Multivariate Linear Mixed Model fit by REML         
## **********************  sommer 4.0  ********************** 
## ============================================================
##          logLik      AIC      BIC Method Converge
## Value -165.2286 332.4571 336.4486     NR     TRUE
## ============================================================
## Variance-Covariance components:
##                     VarComp VarCompSE Zratio Constraint
## 3:GCA2.Yield-Yield    26.64     26.16 1.0185   Positive
## 4:GCA2.Yield-Yield    37.51     37.78 0.9927   Positive
## 1:units.Yield-Yield  216.77     30.75 7.0489   Positive
## 2:units.Yield-Yield  216.77     30.75 7.0489   Positive
## 3:units.Yield-Yield  503.62     77.87 6.4673   Positive
## 4:units.Yield-Yield  738.86    114.17 6.4715   Positive
## ============================================================
## Fixed effects:
##   Trait      Effect Estimate Std.Error t.value
## 1 Yield (Intercept)    138.1    0.9147     151
## ============================================================
## Groups and observations:
##        Yield
## 3:GCA2    20
## 4:GCA2    20
## ============================================================
## Use the &#39;$&#39; sign to access results and parameters
</code></pre>

<h2>3) Using the pin calculator</h2>

<p>Sometimes the user needs to calculate ratios or functions of specific variance-covariance components and obtain the standard error for such parameters. Examples of these are the genetic correlations, heritabilities, etc. Using the CPdata we will show how to estimate the heritability and the standard error using the pin function that uses the delta method to come up with these parameters. This can be extended for any linear combination of the variance components.</p>

<pre><code class="r">data(DT_cpdata)
DT &lt;- DT_cpdata
GT &lt;- GT_cpdata
MP &lt;- MP_cpdata
### look at the data
A &lt;- A.mat(GT) # additive relationship matrix
ans &lt;- mmer(color~1, 
                random=~vs(id,Gu=A), 
                rcov=~units,
                data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -137.304   21:49:54      0           0
##     2      -115.507   21:49:54      0           0
##     3      -111.236   21:49:54      0           0
##     4      -110.755   21:49:54      0           0
##     5      -110.741   21:49:54      0           0
##     6      -110.741   21:49:54      0           0
</code></pre>

<pre><code class="r">(summary(ans.ADE)$varcomp)
</code></pre>

<pre><code>##                       VarComp    VarCompSE   Zratio Constraint
## u:id.color-color  0.003662202 0.0012194130 3.003250   Positive
## u:idd.color-color 0.001820079 0.0007406216 2.457502   Positive
## units.color-color 0.002106929 0.0002864724 7.354736   Positive
</code></pre>

<pre><code class="r">pin(ans, h2 ~ (V1) / ( V1+V2) )
</code></pre>

<pre><code>##     Estimate         SE
## h2 0.6512157 0.06107574
</code></pre>

<p>The same can be used for multivariate models. Please check the documentation of the <code>pin</code> function to see more examples.</p>

<h2>4) Half and full diallel designs (use of the overlay)</h2>

<p>When breeders are looking for the best single cross combinations, diallel designs have been by far the most used design in crops like maize. There are 4 types of diallel designs depending if reciprocate and self cross (omission of parents) are performed (full diallel with parents n<sup>2;</sup> full diallel without parents n(n-1); half diallel with parents &frac12; * n(n+1); half diallel without parents &frac12; * n(n-1) ). In this example we will show a full dialle design (reciprocate crosses are performed) and half diallel designs (only one of the directions is performed).</p>

<p>In the first data set we show a full diallel among 40 lines from 2 heterotic groups, 20 in each. Therefore 400 possible hybrids are possible. We have pehnotypic data for 100 of them across 4 locations. We use the data available to fit a model of the form:</p>

<p><br></p>

<p>\(y = X\beta + Zu_1 + Zu_2 + Zu_S + \epsilon\) </p>

<p><br></p>

<p>We estimate variance components for \(GCA_1\), \(GCA_2\) and \(SCA\) and use them to estimate heritability. Additionally BLUPs for GCA and SCA effects can be used to predict crosses.</p>

<pre><code class="r">data(DT_cornhybrids)
DT &lt;- DT_cornhybrids
DTi &lt;- DTi_cornhybrids
GT &lt;- GT_cornhybrids

modFD &lt;- mmer(Yield~Location, 
               random=~GCA1+GCA2+SCA, 
               rcov=~units,
               data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -149.436   21:49:54      0           0
##     2      -136.475   21:49:54      0           1
##     3      -132.852   21:49:54      0           1
##     4      -132.625   21:49:54      0           1
##     5      -132.596   21:49:55      1           1
##     6      -132.59   21:49:55      1           1
##     7      -132.589   21:49:55      1           1
##     8      -132.589   21:49:55      1           1
</code></pre>

<pre><code class="r">(suma &lt;- summary(modFD)$varcomp)
</code></pre>

<pre><code>##                      VarComp VarCompSE     Zratio Constraint
## GCA1.Yield-Yield    0.000000  16.50337  0.0000000   Positive
## GCA2.Yield-Yield    7.412226  18.94200  0.3913116   Positive
## SCA.Yield-Yield   187.560303  41.59428  4.5092817   Positive
## units.Yield-Yield 221.142463  18.14716 12.1860656   Positive
</code></pre>

<pre><code class="r">Vgca &lt;- sum(suma[1:2,1])
Vsca &lt;- suma[3,1]
Ve &lt;- suma[4,1]
Va = 4*Vgca
Vd = 4*Vsca
Vg &lt;- Va + Vd
(H2 &lt;- Vg / (Vg + (Ve)) )
</code></pre>

<pre><code>## [1] 0.7790856
</code></pre>

<pre><code class="r">(h2 &lt;- Va / (Vg + (Ve)) )
</code></pre>

<pre><code>## [1] 0.02961832
</code></pre>

<p>Don&#39;t worry too much about the small h2 value, the data was simulated to be mainly dominance variance, therefore the Va was simulated extremely small leading to such value of narrow sense h2.</p>

<p>In this second data set we show a small half diallel with 7 parents crossed in one direction. n(n-1)/2 crosses are possible 7(6)/2 = 21 unique crosses. Parents appear as males or females indistictly. Each with two replications in a CRD. For a half diallel design a single GCA variance component for both males and females can be estimated and an SCA as well (\(\sigma^2_GCA\) and \(\sigma^2_SCA\) respectively), and BLUPs for GCA and SCA of the parents can be extracted. We would show first how to use it with the <code>mmer</code> function using the <code>overlay()</code> function. The specific model here is:</p>

<p>\(y = X\beta + Zu_g + Zu_s + \epsilon\) </p>

<pre><code class="r">data(&quot;DT_halfdiallel&quot;)
DT &lt;- DT_halfdiallel
head(DT)
</code></pre>

<pre><code>##   rep geno male female     sugar
## 1   1   12    1      2 13.950509
## 2   2   12    1      2  9.756918
## 3   1   13    1      3 13.906355
## 4   2   13    1      3  9.119455
## 5   1   14    1      4  5.174483
## 6   2   14    1      4  8.452221
</code></pre>

<pre><code class="r">DT$femalef &lt;- as.factor(DT$female)
DT$malef &lt;- as.factor(DT$male)
DT$genof &lt;- as.factor(DT$geno)
#### model using overlay
modh &lt;- mmer(sugar~1, 
             random=~vs(overlay(femalef,malef)) 
             + genof,
             data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -10.425   21:49:55      0           0
##     2      -6.487   21:49:55      0           0
##     3      -5.732   21:49:55      0           0
##     4      -5.67494   21:49:55      0           0
##     5      -5.67441   21:49:55      0           0
</code></pre>

<pre><code class="r">summary(modh)$varcomp
</code></pre>

<pre><code>##                        VarComp VarCompSE   Zratio Constraint
## u:femalef.sugar-sugar 5.507899 3.5741151 1.541052   Positive
## genof.sugar-sugar     1.815784 1.3629575 1.332238   Positive
## units.sugar-sugar     3.117538 0.9626094 3.238632   Positive
</code></pre>

<p>Notice how the <code>overlay()</code> argument makes the overlap of incidence matrices possible making sure that male and female are joint into a single random effect.</p>

<h2>5) Genomic selection</h2>

<p>In this section we will use wheat data from CIMMYT to show how is genomic selection performed. This is the case of prediction of specific individuals within a population. It basically uses a similar model of the form:</p>

<p><br></p>

<p>\(y = X\beta + Zu + \epsilon\) </p>

<p><br></p>

<p>and takes advantage of the variance covariance matrix for the genotype effect known as the additive relationship matrix (A) and calculated using the <code>A.mat</code> function to establish connections among all individuals and predict the BLUPs for individuals that were not measured. The prediction accuracy depends on several factors such as the heritability (\(h^2\)), training population used (TP), size of TP, etc.</p>

<pre><code class="r">data(DT_wheat)
DT &lt;- DT_wheat
GT &lt;- GT_wheat
colnames(DT) &lt;- paste0(&quot;X&quot;,1:ncol(DT))
DT &lt;- as.data.frame(DT);DT$id &lt;- as.factor(rownames(DT))
# select environment 1
rownames(GT) &lt;- rownames(DT)
K &lt;- A.mat(GT) # additive relationship matrix
colnames(K) &lt;- rownames(K) &lt;- rownames(DT)
# GBLUP pedigree-based approach
set.seed(12345)
y.trn &lt;- DT
vv &lt;- sample(rownames(DT),round(nrow(DT)/5))
y.trn[vv,&quot;X1&quot;] &lt;- NA
head(y.trn)
</code></pre>

<pre><code>##              X1          X2          X3         X4   id
## 775   1.6716295 -1.72746986 -1.89028479  0.0509159  775
## 2166 -0.2527028  0.40952243  0.30938553 -1.7387588 2166
## 2167         NA -0.64862633 -0.79955921 -1.0535691 2167
## 2465  0.7854395  0.09394919  0.57046773  0.5517574 2465
## 3881  0.9983176 -0.28248062  1.61868192 -0.1142848 3881
## 3889  2.3360969  0.62647587  0.07353311  0.7195856 3889
</code></pre>

<pre><code class="r">## GBLUP
ans &lt;- mmer(X1~1,
            random=~vs(id,Gu=K), 
            rcov=~units, 
            data=y.trn) # kinship based
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -207.28   21:49:55      0           0
##     2      -204.823   21:49:55      0           0
##     3      -203.988   21:49:55      0           0
##     4      -203.866   21:49:56      1           0
##     5      -203.862   21:49:56      1           0
##     6      -203.861   21:49:56      1           0
</code></pre>

<pre><code class="r">ans$U$`u:id`$X1 &lt;- as.data.frame(ans$U$`u:id`$X1)
rownames(ans$U$`u:id`$X1) &lt;- gsub(&quot;id&quot;,&quot;&quot;,rownames(ans$U$`u:id`$X1))
cor(ans$U$`u:id`$X1[vv,],DT[vv,&quot;X1&quot;], use=&quot;complete&quot;)
</code></pre>

<pre><code>## [1] 0.5737594
</code></pre>

<pre><code class="r">## rrBLUP
ans2 &lt;- mmer(X1~1,
             random=~vs(list(GT)), 
             rcov=~units,
             data=y.trn) # kinship based
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -351.655   21:49:57      1           0
##     2      -249.864   21:49:57      1           0
##     3      -214.384   21:49:57      1           0
##     4      -204.423   21:49:57      1           0
##     5      -203.883   21:49:57      1           0
##     6      -203.862   21:49:58      2           0
##     7      -203.861   21:49:58      2           0
</code></pre>

<pre><code class="r">u &lt;- GT %*% as.matrix(ans2$U$`u:GT`$X1) # BLUPs for individuals
rownames(u) &lt;- rownames(GT)
cor(u[vv,],DT[vv,&quot;X1&quot;]) # same correlation
</code></pre>

<pre><code>## [1] 0.5737681
</code></pre>

<pre><code class="r"># the same can be applied in multi-response models in GBLUP or rrBLUP
</code></pre>

<h2>6) Single cross prediction</h2>

<p>When doing prediction of single cross performance the phenotype can be dissected in three main components, the general combining abilities (GCA) and specific combining abilities (SCA). This can be expressed with the same model analyzed in the diallel experiment mentioned before:</p>

<p><br></p>

<p>\(y = X\beta + Zu_1 + Zu_2 + Zu_S + \epsilon\) </p>

<p><br></p>

<p>with:</p>

<p><br></p>

<p>\(u_1\) ~ N(0, \(K_1\)\(\sigma^2_u1\))</p>

<p>\(u_2\) ~ N(0, \(K_2\)\(\sigma^2_u2\))</p>

<p>\(u_s\) ~ N(0, \(K_3\)\(\sigma^2_us\))</p>

<p><br></p>

<p>And we can specify the K matrices. The main difference between this model and the full and half diallel designs is the fact that this model will include variance covariance structures in each of the three random effects (GCA1, GCA2 and SCA) to be able to predict the crosses that have not ocurred yet. We will use the data published by Technow et al. (2015) to show how to do prediction of single crosses.</p>

<pre><code class="r">data(DT_technow)
DT &lt;- DT_technow
Md &lt;- Md_technow
Mf &lt;- Mf_technow
Ad &lt;- Ad_technow
Af &lt;- Af_technow
# RUN THE PREDICTION MODEL
y.trn &lt;- DT
vv1 &lt;- which(!is.na(DT$GY))
vv2 &lt;- sample(vv1, 100)
y.trn[vv2,&quot;GY&quot;] &lt;- NA
anss2 &lt;- mmer(GY~1, 
               random=~vs(dent,Gu=Ad) + vs(flint,Gu=Af), 
               rcov=~units,
               data=y.trn) 
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      104.311   21:49:59      1           0
##     2      143.766   21:50:0      2           0
##     3      152.662   21:50:1      3           0
##     4      153.91   21:50:1      3           0
##     5      153.974   21:50:2      4           0
##     6      153.978   21:50:3      5           0
##     7      153.978   21:50:4      6           0
</code></pre>

<pre><code class="r">summary(anss2)$varcomp
</code></pre>

<pre><code>##                VarComp VarCompSE    Zratio Constraint
## u:dent.GY-GY  16.06423 2.5737578  6.241548   Positive
## u:flint.GY-GY 11.42070 2.1591718  5.289390   Positive
## units.GY-GY   16.81801 0.7689509 21.871368   Positive
</code></pre>

<pre><code class="r">zu1 &lt;- model.matrix(~dent-1,y.trn) %*% anss2$U$`u:dent`$GY
zu2 &lt;- model.matrix(~flint-1,y.trn) %*% anss2$U$`u:flint`$GY
u &lt;- zu1+zu2+anss2$Beta[1,&quot;Estimate&quot;]
cor(u[vv2,], DT$GY[vv2])
</code></pre>

<pre><code>## [1] 0.7756383
</code></pre>

<p>In the previous model we only used the GCA effects (GCA1 and GCA2) for practicity, altough it&#39;s been shown that the SCA effect doesn&#39;t actually help that much in increasing prediction accuracy and increase a lot the computation intensity required since the variance covariance matrix for SCA is the kronecker product of the variance covariance matrices for the GCA effects, resulting in a 10578x10578 matrix that increases in a very intensive manner the computation required.</p>

<p>A model without covariance structures would show that the SCA variance component is insignificant compared to the GCA effects. This is why including the third random effect doesn&#39;t increase the prediction accuracy.</p>

<h2>7) Spatial modeling (using the 2-dimensional spline)</h2>

<p>We will use the CPdata to show the use of 2-dimensional splines for accomodating spatial effects in field experiments. In early generation variety trials the availability of seed is low, which makes the use of unreplicated design a neccesity more than anything else. Experimental designs such as augmented designs and partially-replicated (p-rep) designs become every day more common this days. </p>

<p>In order to do a good job modeling the spatial trends happening in the field special covariance structures have been proposed to accomodate such spatial trends (i.e. autoregressive residuals; ar1). Unfortunately, some of these covariance structures make the modeling rather unstable. More recently other research groups have proposed the use of 2-dimensional splines to overcome such issues and have a more robust modeling of the spatial terms (Lee et al. 2013; Rodríguez-Álvarez et al. 2018).</p>

<p>In this example we assume an unreplicated population where row and range information is available which allows us to fit a 2 dimensional spline model.</p>

<pre><code class="r">data(DT_cpdata)
DT &lt;- DT_cpdata
GT &lt;- GT_cpdata
MP &lt;- MP_cpdata
### mimic two fields
A &lt;- A.mat(GT)
mix &lt;- mmer(Yield~1,
            random=~vs(id, Gu=A) +
              vs(Rowf) +
              vs(Colf) +
              vs(spl2D(Row,Col)),
            rcov=~vs(units),
            data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -154.198   21:50:4      0           0
##     2      -152.064   21:50:4      0           0
##     3      -151.265   21:50:4      0           0
##     4      -151.202   21:50:4      0           0
##     5      -151.201   21:50:5      1           0
</code></pre>

<pre><code class="r">summary(mix)
</code></pre>

<pre><code>## ============================================================
##          Multivariate Linear Mixed Model fit by REML         
## **********************  sommer 4.0  ********************** 
## ============================================================
##          logLik      AIC      BIC Method Converge
## Value -151.2011 304.4021 308.2938     NR     TRUE
## ============================================================
## Variance-Covariance components:
##                     VarComp VarCompSE Zratio Constraint
## u:id.Yield-Yield      783.4     319.3 2.4536   Positive
## u:Rowf.Yield-Yield    814.7     390.5 2.0863   Positive
## u:Colf.Yield-Yield    182.2     129.7 1.4053   Positive
## u:Row.Yield-Yield     513.6     694.7 0.7393   Positive
## u:units.Yield-Yield  2922.6     294.1 9.9368   Positive
## ============================================================
## Fixed effects:
##   Trait      Effect Estimate Std.Error t.value
## 1 Yield (Intercept)    132.1     8.791   15.03
## ============================================================
## Groups and observations:
##        Yield
## u:id     363
## u:Rowf    13
## u:Colf    36
## u:Row    168
## ============================================================
## Use the &#39;$&#39; sign to access results and parameters
</code></pre>

<p>Notice that the job is done by the <code>spl2D()</code> function that takes the Row and Col information to fit a spatial kernel. </p>

<h2>8) Multivariate genetic models and genetic correlations</h2>

<p>Sometimes is important to estimate genetic variance-covariance among traits, multi-reponse models are very useful for such task. Let see an example with 3 traits (color, Yield, and Firmness) and a single random effect (genotype; id) although multiple effects can be modeled as well. We need to use a variance covariance structure for the random effect to be able to obtain the genetic covariance among traits.</p>

<pre><code class="r">data(DT_cpdata)
DT &lt;- DT_cpdata
GT &lt;- GT_cpdata
MP &lt;- MP_cpdata
A &lt;- A.mat(GT)
ans.m &lt;- mmer(cbind(Yield,color)~1,
               random=~ vs(id, Gu=A, Gtc=unsm(2))
               + vs(Rowf,Gtc=diag(2))
               + vs(Colf,Gtc=diag(2)),
               rcov=~ vs(units, Gtc=unsm(2)),
               data=DT)
</code></pre>

<pre><code>## iteration    LogLik     wall    cpu(sec)   restrained
##     1      -375.872   21:50:6      1           0
##     2      -291.932   21:50:6      1           0
##     3      -258.273   21:50:7      2           0
##     4      -253.459   21:50:7      2           0
##     5      -253.291   21:50:7      2           0
##     6      -253.278   21:50:8      3           0
##     7      -253.277   21:50:8      3           0
##     8      -253.277   21:50:9      4           0
</code></pre>

<p>Now you can extract the BLUPs using the &#39;randef&#39; function or simple accesing with the &#39;$&#39; sign and pick &#39;u.hat&#39;. Also, genetic correlations and heritabilities can be calculated easily. </p>

<pre><code class="r">cov2cor(ans.m$sigma$`u:id`)
</code></pre>

<pre><code>##           Yield     color
## Yield 1.0000000 0.1234441
## color 0.1234441 1.0000000
</code></pre>

<h2>9) Final remarks</h2>

<p>Keep in mind that sommer uses direct inversion (DI) algorithm which can be very slow for large datasets. The package is focused in problems of the type p &gt; n (more random effect levels than observations) and models with dense covariance structures. For example, for experiment with dense covariance structures with low-replication (i.e. 2000 records from 1000 individuals replicated  twice with a covariance structure of 1000x1000) sommer will be faster than MME-based software. Also for genomic problems with large number of random effect levels, i.e. 300 individuals (n) with 100,000 genetic markers (p). For highly replicated trials with small covariance structures or n &gt; p (i.e. 2000 records from 200 individuals replicated 10 times with covariance structure of 200x200) asreml or other MME-based algorithms will be much faster and we recommend you to opt for those software.</p>

<h2>Literature</h2>

<p>Covarrubias-Pazaran G. 2016. Genome assisted prediction of quantitative traits using the R package sommer. PLoS ONE 11(6):1-15.</p>

<p>Covarrubias-Pazaran G. 2018. Software update: Moving the R package sommer to multivariate mixed models for genome-assisted prediction. doi: <a href="https://doi.org/10.1101/354639">https://doi.org/10.1101/354639</a></p>

<p>Bernardo Rex. 2010. Breeding for quantitative traits in plants. Second edition. Stemma Press. 390 pp.</p>

<p>Gilmour et al. 1995. Average Information REML: An efficient algorithm for variance parameter estimation in linear mixed models. Biometrics 51(4):1440-1450.</p>

<p>Henderson C.R. 1975. Best Linear Unbiased Estimation and Prediction under a Selection Model. Biometrics vol. 31(2):423-447.</p>

<p>Kang et al. 2008. Efficient control of population structure in model organism association mapping. Genetics 178:1709-1723.</p>

<p>Lee, D.-J., Durban, M., and Eilers, P.H.C. (2013). Efficient two-dimensional smoothing with P-spline ANOVA mixed models and nested bases. Computational Statistics and Data Analysis, 61, 22 - 37.</p>

<p>Lee et al. 2015. MTG2: An efficient algorithm for multivariate linear mixed model analysis based on genomic information. Cold Spring Harbor. doi: <a href="http://dx.doi.org/10.1101/027201">http://dx.doi.org/10.1101/027201</a>.</p>

<p>Maier et al. 2015. Joint analysis of psychiatric disorders increases accuracy of risk prediction for schizophrenia, bipolar disorder, and major depressive disorder. Am J Hum Genet; 96(2):283-294.</p>

<p>Rodriguez-Alvarez, Maria Xose, et al. Correcting for spatial heterogeneity in plant breeding experiments with P-splines. Spatial Statistics 23 (2018): 52-71.</p>

<p>Searle. 1993. Applying the EM algorithm to calculating ML and REML estimates of variance components. Paper invited for the 1993 American Statistical Association Meeting, San Francisco.</p>

<p>Yu et al. 2006. A unified mixed-model method for association mapping that accounts for multiple levels of relatedness. Genetics 38:203-208.</p>

<p>Abdollahi Arpanahi R, Morota G, Valente BD, Kranis A, Rosa GJM, Gianola D. 2015. Assessment of bagging GBLUP for whole genome prediction of broiler chicken traits. Journal of Animal Breeding and Genetics 132:218-228.</p>

<p>Tunnicliffe W. 1989. On the use of marginal likelihood in time series model estimation. JRSS 51(1):15-27.</p>

</body>

</html>
